## map(function, iterable)
Functionality:
Applies a function to every item in an iterable (like a list) and returns a new iterable with the results.

### Simple words:
Transforming each item.

In [7]:
l = [1,2,3,4,5,6]

In [9]:
[i**2 for i in l]

[1, 4, 9, 16, 25, 36]

In [11]:
# 1. Square each number
numbers = [1, 2, 3, 4]
(lambda x: x**2, numbers)

(<function __main__.<lambda>(x)>, [1, 2, 3, 4])

In [10]:
list(map(lambda x: x**2 , numbers))

[1, 4, 9, 16]

In [12]:
squares = list(map(lambda x: x**2, numbers))

In [14]:
squares 

[1, 4, 9, 16]

In [16]:
list(map(lambda x: x**3 , numbers))

[1, 8, 27, 64]

In [13]:
def str(n):
    return n.upper()

In [15]:
str("chakresh")

'CHAKRESH'

In [3]:
# 2. Convert to uppercase
words = ["hello", "world", "chakresh", "unnisa", "sekhar", "keziya"]
uppercased = list(map(str.upper, words))

In [5]:
uppercased

['HELLO', 'WORLD', 'CHAKRESH', 'UNNISA', 'SEKHAR', 'KEZIYA']

In [7]:
lowercased = list(map(str.lower, words))

In [9]:
lowercased

['hello', 'world', 'chakresh', 'unnisa', 'sekhar', 'keziya']

In [11]:
# 3. Celsius to Fahrenheit
celsius = [0, 10, 20, 30]
fahrenheit = list(map(lambda c: (c * 9/5) + 32, celsius))

In [13]:
fahrenheit

[32.0, 50.0, 68.0, 86.0]

# 4. Add index to value

In [22]:
values = [10, 20, 30]

In [24]:
enumerate(values)

In [26]:
with_index = list(map(lambda i_x: i_x[0] + i_x[1], enumerate(values)))

In [28]:
with_index

[10, 21, 32]

In [30]:
# 5. Add items from two lists
a = [1, 2, 3]
b = [4, 5, 6]
sum = list(map(lambda x, y: x + y, a, b))

In [32]:
sum

[5, 7, 9]

In [34]:
# 5. Add items from three lists
a = [1, 2, 3]
b = [4, 5, 6]
c = [6,7,8]
sum = list(map(lambda x, y,z : x + y +z, a, b, c))

In [38]:
a = [1, 2, 3]
b = [4, 5, 6]
c = [6,7,8]

prd = list(map(lambda x,y,z : x*y*z, a,b,c))

In [40]:
prd

[24, 70, 144]

In [36]:
sum

[11, 14, 17]

### filter(function, iterable)
Functionality:
Keeps only items from an iterable where the function returns True. It filters out the rest.

Pick items based on a condition.

In [45]:
def even(l):
    for i in l:
        if i%2==0:
            print(i)

In [47]:
even([1,2,3,4,5,6,7])

2
4
6


In [51]:
# 1. Keep even numbers
numbers = [1, 2, 3, 4, 5]
evens = list(filter(lambda x: x % 2 == 0, numbers))

In [53]:
print(evens)

[2, 4]


In [55]:
# 2. Remove empty strings
strings = ["hello", "", "world", ""]
non_empty = list(filter(None, strings))

In [57]:
non_empty

['hello', 'world']

In [59]:
def test(l):
    for i in l:
        if i>10:
            print(i)

In [63]:
test([10,12,4,5,6,])

12


In [65]:
# 3. Numbers greater than 10
numbers = [5, 12, 17, 3, 9]
numbers_greaterthan_10 = list(filter(lambda x: x > 10, numbers))

In [69]:
numbers_greaterthan_10

[12, 17]

In [71]:
def palindrom(l):
    l1 = []
    for i in l:
        a = i[::-1]
        if a == i:
            l1.append(i)
    return l1

In [73]:
palindrom(["madam", "racecar", "apple", "level"])

['madam', 'racecar', 'level']

In [3]:
s = "madam"

def test(s):
    return s == s[::-1]

In [5]:
test(s)

True

In [7]:
def palind(i):
    a = i[::-1]
    if a == i:
        print("it is a palindrom")
    else:
        print("it's not a palindrom")

In [9]:
palind("aimnxt")

it's not a palindrom


In [11]:
palind("level")

it is a palindrom


In [83]:
def palin(s):
    a = s[::-1]
    if s==a:
        print("given string is palindrom")
    else:
        print("given string is not palindrom")

In [93]:
palin("nitin")

given string is palindrom


In [13]:
l = ["madam", "level", "apple", "racecar"]
palindromes = list(filter(lambda x : x == x[::-1], l))

In [15]:
palindromes

['madam', 'level', 'racecar']

In [95]:
# 4. Keep palindromes
words = ["madam", "racecar", "apple", "level"]
palindromes = list(filter(lambda w: w == w[::-1], words))

In [97]:
palindromes

['madam', 'racecar', 'level']

In [99]:
# 5. Keep truthy values
data = [0, "", None, "hello", 42]
truthy = list(filter(bool, data))

In [101]:
truthy

['hello', 42]

### reduce(function, iterable)
Functionality:
Takes a list and reduces it to a single value, by applying a function cumulatively to the items.

Roll everything into one value

In [37]:
from functools import reduce

In [31]:
def test(l):
    sum = 0
    for i in l:
        if type(i) ==int:
            sum+=i
    print(sum)

In [33]:
test([1,2,4,5])

12


In [39]:
# 1. Sum of numbers
nums = [1, 2, 3, 4]
total = reduce(lambda x, y: x + y, nums)

In [41]:
total

10

In [43]:
# 2. Multiply all numbers
nums = [2, 3, 4]
product = reduce(lambda x, y: x * y, nums)

In [45]:
# 3. Find the max
nums = [10, 20, 5, 40]
maximum = reduce(lambda x, y: x if x > y else y, nums)

In [47]:
maximum

40

In [49]:
# 4. Concatenate strings
words = ["Python", "is", "fun"]
sentence = reduce(lambda x, y: x + " " + y, words)

In [51]:
sentence

'Python is fun'

In [53]:
# 5. Make number from digits
digits = [1, 2, 3, 4]
number = reduce(lambda x, y: x * 10 + y, digits)

In [55]:
number

1234

In [ ]:
# 5. Make number from digits
digits = [1, -2, 3, 4]
number = reduce(lambda x, y: x * 10 + y, digits)

Given a list of numbers, return the sum of squares of only the odd numbers.

In [67]:
from functools import reduce
numbers = [1, 2, 3, 4, 5, 6]
# Step 1: Filter odd numbers
odd = list(filter(lambda x: x % 2 != 0, numbers))  # [1, 3, 5]

# Step 2: Multiply them using reduce
product = reduce(lambda x, y: x * y, odd)

print(product)  # Output: 48


15


In [69]:
from functools import reduce
numbers = [1, 2, 3, 4, 5, 6]
# Step 1: Filter odd numbers
odd = list(filter(lambda x: x % 2 != 0, numbers))  # [1, 3, 5]

## Square the odd numbers using Map function
squares = list(map(lambda x: x**2, odd))

# Step 2: Multiply them using reduce
product = reduce(lambda x, y: x * y, squares)

print(product)  # Output: 48


225


In [71]:
odd

[1, 3, 5]

In [73]:
squares

[1, 9, 25]

In [ ]:
from functools import reduce

numbers = [1, 2, 3, 4, 5, 6]

product = reduce(lambda x, y: x * y, filter(lambda x: x % 2 == 0, numbers))

print(product)  # Output: 48


Practical Questions on map()
1. Convert a list of strings to uppercase

words = ['python', 'is', 'awesome']
Output: ['PYTHON', 'IS', 'AWESOME']

2. Square all numbers in a list

numbers = [1, 2, 3, 4]
utput: [1, 4, 9, 16]

3. Convert a list of integers to strings

nums = [1, 2, 3]
Output: ['1', '2', '3']

4. Add 5 to each element in a list
marks = [50, 60, 70]
Output: [55, 65, 75]

5. Get lengths of all strings in a list

names = ['apple', 'banana', 'kiwi']
Output: [5, 6, 4]


1. Filter out even numbers

nums = [1, 2, 3, 4, 5, 6]
Output: [2, 4, 6]

2. Keep only strings with more than 3 characters

words = ['hi', 'hello', 'hey', 'sun']
Output: ['hello']

3. Remove empty strings from a list

data = ['Python', '', 'Java', '', 'C++']
Output: ['Python', 'Java', 'C++']

4. Get names that start with 'A'
names = ['Alice', 'Bob', 'Angela', 'Tom']
Output: ['Alice', 'Angela']

5. Keep truthy values only

items = [0, None, 'hello', False, 42, '']
Output: ['hello', 42]

1. Sum of all numbers

nums = [1, 2, 3, 4]
Output: 10

2. Product of all even numbers
nums = [1, 2, 3, 4, 5]
Filter even → [2, 4]
Reduce → 8

3. Find the longest word
words = ['apple', 'banana', 'kiwi']
Output: 'banana'

4. Join list of strings into one sentence
words = ['Python', 'is', 'fun']
Output: 'Python is fun'

5. Create number from digits

digits = [1, 2, 3, 4]
Output: 1234